device/pcie/pci_device.cpp

```diff
@@ -297,63 +309,64 @@ void PCIDevice::setup_device() {
-            throw std::runtime_error(fmt::format("BAR4 UC mapping failed for device {}.", device_id));
+            throw std::runtime_error(fmt::format("BAR4 UC mapping failed for device {}.", device_num));
         }
 
         system_reg_start_offset = (512 - 16) * 1024*1024;
         system_reg_offset_adjust = (512 - 32) * 1024*1024;
     } else if(arch == tt::ARCH::BLACKHOLE) {
         if (bar2_uc_mapping.mapping_id != TENSTORRENT_MAPPING_RESOURCE1_UC) {
-            throw std::runtime_error(fmt::format("Device {} has no BAR2 UC mapping.", device_id));
+            throw std::runtime_error(fmt::format("Device {} has no BAR2 UC mapping.", device_num));
         }
 
         // Using UnCachable memory mode. This is used for accessing registers on Blackhole.
         bar2_uc_size = bar2_uc_mapping.mapping_size;
-        bar2_uc = mmap(NULL, bar2_uc_mapping.mapping_size, PROT_READ | PROT_WRITE, MAP_SHARED, device_fd, bar2_uc_mapping.mapping_base);
+        bar2_uc = mmap(NULL, bar2_uc_mapping.mapping_size, PROT_READ | PROT_WRITE, MAP_SHARED, fd, bar2_uc_mapping.mapping_base);
 
         if (bar2_uc == MAP_FAILED) {
-            throw std::runtime_error(fmt::format("BAR2 UC mapping failed for device {}.", device_id));
+            throw std::runtime_error(fmt::format("BAR2 UC mapping failed for device {}.", device_num));
         }
 
         if (bar4_wc_mapping.mapping_id != TENSTORRENT_MAPPING_RESOURCE2_WC) {
-            throw std::runtime_error(fmt::format("Device {} has no BAR4 WC mapping.", device_id));
+            throw std::runtime_error(fmt::format("Device {} has no BAR4 WC mapping.", device_num));
         }
 
         // Using Write-Combine memory mode. This is used for accessing DRAM on Blackhole.
         // WC doesn't guarantee write ordering but has better performance.
         bar4_wc_size = bar4_wc_mapping.mapping_size;
-        bar4_wc = mmap(NULL, bar4_wc_mapping.mapping_size, PROT_READ | PROT_WRITE, MAP_SHARED, device_fd, bar4_wc_mapping.mapping_base);
+        bar4_wc = mmap(NULL, bar4_wc_mapping.mapping_size, PROT_READ | PROT_WRITE, MAP_SHARED, fd, bar4_wc_mapping.mapping_base);
 
         if (bar4_wc == MAP_FAILED) {
-            throw std::runtime_error(fmt::format("BAR4 WC mapping failed for device {}.", device_id));
+            throw std::runtime_error(fmt::format("BAR4 WC mapping failed for device {}.", device_num));
         }
     }
 
     // GS+WH: ARC_SCRATCH[6], BH: NOC NODE_ID
     read_checking_offset = arch == tt::ARCH::BLACKHOLE ? BH_NOC_NODE_ID_OFFSET : GS_WH_ARC_SCRATCH_6_OFFSET;
 }
 
-void PCIDevice::close_device() {
+PCIDevice::~PCIDevice() {
     if (arch == tt::ARCH::BLACKHOLE && bar2_uc != nullptr && bar2_uc != MAP_FAILED) {
         // Disable ATU index 0
         // TODO: Implement disabling for all indexes, once more host channels are enabled.
+
+        // This is not going to happen if the application crashes, so if it's
+        // essential for correctness then it needs to move to the driver.
```

Comment:
Yeah, if we deleted this code, everything would still work. I was just trying to be diligent, but this code is not necessary (since you setup the ATU each time you start using the device). 
Just FYI, no action on this comment.